File: src-lang/js/react/view/backend.clj
(ns js.react.view.backend
  "Native React implementations for substrate view component ids."
  (:require [hara.lang :as l]))

(l/script :js
  {:require [[xt.lang.spec-base :as xt]
             [xt.lang.common-data :as xtd]
             [js.react :as r]]})

(defn.js native-registry
  []
  (return
   {"ui/fragment" {"tag" r/Fragment}
    "ui/row" {"tag" "div" "layout" "row"}
    "ui/column" {"tag" "div" "layout" "column"}
    "ui/text" {"tag" "span" "value_prop" "value"}
    "ui/title" {"tag" "h2" "value_prop" "value"}
    "ui/label" {"tag" "label" "value_prop" "value"}
    "ui/input" {"tag" "input" "input" true}
    "ui/button" {"tag" "button" "press" true}
    "ui/alert" {"tag" "div" "layout" "column"}
    "ui/spinner" {"tag" "span" "value" "Loading…"}}))

(defn.js registry
  [overrides polyfills]
  (return {"backend" "react"
           "native" (-/native-registry)
           "polyfills" (or polyfills {})
           "overrides" (or overrides {})}))

(defn.js style
  "maps portable layout and token props to React style"
  [props entry]
  (var out (xt/x:obj-clone (or (xt/x:get-key props "style") {})))
  (var layout (xt/x:get-key entry "layout"))
  (when layout
    (xt/x:set-key out "display" "flex")
    (xt/x:set-key out "flexDirection" layout))
  (xt/for:array [pair [["gap" "gap"]
                       ["padding" "padding"]
                       ["align" "alignItems"]
                       ["justify" "justifyContent"]]]
    (var source (xt/x:get-idx pair 0))
    (var target (xt/x:get-idx pair 1))
    (when (xt/x:not-nil? (xt/x:get-key props source))
      (xt/x:set-key out target (xt/x:get-key props source))))
  (var tone (xt/x:get-key props "tone"))
  (when (== tone "danger")
    (xt/x:set-key out "color" "#b91c1c")
    (xt/x:set-key out "backgroundColor" "#fef2f2"))
  (return out))

(defn.js dom-props
  [runtime props entry]
  (var out {})
  (xt/for:array [key ["id" "value" "placeholder" "type" "disabled"
                      "readOnly" "rows" "title" "className"]]
    (when (xt/x:has-key? props key)
      (xt/x:set-key out key (xt/x:get-key props key))))
  (var class-name (or (xt/x:get-key props "class_name")
                      (xt/x:get-key props "class")))
  (when class-name
    (xt/x:set-key out "className" class-name))
  (var aria-label (xt/x:get-key props "aria_label"))
  (when aria-label
    (xt/x:set-key out "aria-label" aria-label))
  (var style (-/style props entry))
  (when (> (xt/x:len (xt/x:obj-keys style)) 0)
    (xt/x:set-key out "style" style))
  (var on-change (xt/x:get-key props "on_change"))
  (when on-change
    (xt/x:set-key out "onChange"
                  (fn [event]
                    (return ((xt/x:get-key runtime "dispatch")
                             on-change
                             {"value" (. event target value)})))))
  (var on-press (xt/x:get-key props "on_press"))
  (when on-press
    (xt/x:set-key out "onClick"
                  (fn [event]
                    (return ((xt/x:get-key runtime "dispatch")
                             on-press event)))))
  (return out))

(defn.js render-native
  [runtime entry props children]
  (var tag (xt/x:get-key entry "tag"))
  (var dom-props (-/dom-props runtime props entry))
  (var value-prop (xt/x:get-key entry "value_prop"))
  (var value (:? value-prop (xt/x:get-key props value-prop)
                 (xt/x:get-key entry "value")))
  (return (r/createElement tag dom-props (or value children))))

File: packages-gen/js/libs/xt-ui/react/view/backend.js
const FigmaUi = require("@xtalk/figma-ui")

const React = require("react")

const catalog = require("@xtalk/substrate/view-catalog.js")

const xts = require("@xtalk/lang/common-string.js")

function native_registry(){
  return {
    "ui/description":{"tag":"p","value_prop":"value"},
    "ui/column":{"tag":"div","column":true},
    "ui/card-content":{"tag":FigmaUi.CardContent},
    "ui/row":{"tag":"div","row":true},
    "ui/textarea":{"tag":FigmaUi.Textarea},
    "ui/card-description":{"tag":FigmaUi.CardDescription,"value_prop":"value"},
    "ui/icon":{"tag":"span","value_prop":"value"},
    "ui/text":{"tag":"span","value_prop":"value"},
    "ui/separator":{"tag":FigmaUi.Separator},
    "ui/label":{"tag":FigmaUi.Label,"value_prop":"value"},
    "ui/input":{"tag":FigmaUi.Input},
    "ui/badge":{"tag":FigmaUi.Badge,"value_prop":"value"},
    "ui/card-title":{"tag":FigmaUi.CardTitle,"value_prop":"value"},
    "ui/card-footer":{"tag":FigmaUi.CardFooter},
    "ui/table-row":{"tag":FigmaUi.TableRow},
    "ui/spinner":{"tag":FigmaUi.Skeleton},
    "ui/table":{"tag":FigmaUi.Table},
    "ui/title":{"tag":"h2","value_prop":"value"},
    "ui/table-body":{"tag":FigmaUi.TableBody},
    "ui/card":{"tag":FigmaUi.Card},
    "ui/table-header":{"tag":FigmaUi.TableHeader},
    "ui/table-cell":{"tag":FigmaUi.TableCell,"value_prop":"value"},
    "ui/alert":{"tag":FigmaUi.Alert},
    "ui/fragment":{"tag":React.Fragment},
    "ui/table-head":{"tag":FigmaUi.TableHead,"value_prop":"value"},
    "ui/card-header":{"tag":FigmaUi.CardHeader},
    "ui/button":{"tag":FigmaUi.Button},
    "ui/image":{"tag":"img"},
    "ui/scroll":{"tag":"div"}
  };
}

function pascal_case(s){
  let out = "";
  for(let part of xts.split(s,"-")){
    out = (out + xts.capitalize(part));
  };
  return out;
}

function native_entry(registry,component_id){
  let entry = (registry["native"])[component_id];
  if(null != entry){
    return entry;
  }
  if(catalog.platform_idp(component_id)){
    let name = pascal_case(xts.substring(component_id,3,component_id.length));
    let tag = FigmaUi[name];
    if(null == tag){
      throw "figma component missing [react] - " + component_id;
    }
    return {"tag":tag};
  }
  return null;
}

function registry(overrides,polyfills){
  return {
    "backend":"react",
    "native":native_registry(),
    "polyfills":polyfills || {},
    "overrides":overrides || {}
  };
}

function dom_props(runtime,props,entry){
  let out = {};
  for(let key of [
    "id",
    "value",
    "placeholder",
    "type",
    "disabled",
    "rows",
    "src",
    "alt",
    "variant",
    "size",
    "href"
  ]){
    if(null != props[key]){
      out[key] = props[key];
    }
  };
  if(null != props["for"]){
    out["htmlFor"] = props["for"];
    delete(out["for"]);
  }
  let class_name = props["class"];
  if(entry["row"]){
    class_name = ("flex flex-row " + (class_name || ""));
  }
  if(entry["column"]){
    class_name = ("flex flex-col " + (class_name || ""));
  }
  if(props["pending"]){
    out["disabled"] = true;
    class_name = ((class_name || "") + " opacity-60 pointer-events-none");
  }
  if(class_name){
    out["className"] = class_name;
  }
  let aria_label = props["aria_label"];
  if(aria_label){
    out["aria-label"] = aria_label;
  }
  if(null != props["read_only"]){
    out["readOnly"] = props["read_only"];
  }
  let on_change = props["on_change"];
  if(on_change){
    out["onChange"] = (function (event){
      return runtime["dispatch"](on_change,{"value":event.target.value});
    });
  }
  let on_press = props["on_press"];
  if(on_press){
    out["onClick"] = (function (event){
      return runtime["dispatch"](on_press,event);
    });
  }
  return out;
}

function render_native(runtime,entry,props,children){
  let tag = entry["tag"];
  let dom_props = dom_props(runtime,props,entry);
  let value_prop = entry["value_prop"];
  let value = value_prop ? props[value_prop] : null;
  return React.createElement(tag,dom_props,value || children);
}

module.exports = {
  ["native_registry"]:native_registry,
  ["pascal_case"]:pascal_case,
  ["native_entry"]:native_entry,
  ["registry"]:registry,
  ["dom_props"]:dom_props,
  ["render_native"]:render_native
}